mirage-router/unikernel.ml

```diff
@@ -0,0 +1,271 @@
+   incoming fragment individually. this makes a lot of sense in setups where
+   the >router< may be only one route from the source to destination, and not
+   every ip fragment takes the same route.
+
+   in our specific setup, this is not the case, and it feels more appropriate
+   (better for the network) to reassemble and only forward full ipv4 packets.
+
+   a difference would be either if there's a fault-tolerant setup, or if
+   fragmented packets are used for real-time communication. my intuition is
+   that path mtu discovery or 512 byte sized udp payload have won.
+
+   we strip all ipv4 options (i.e. source route / record route / etc.), but
+   i think ipv4 options are barely used in the wild, so this is fine.
+*)
+
+open Lwt.Infix
+
+module Main (R : Mirage_random.S) (M : Mirage_clock.MCLOCK) (P : Mirage_clock.PCLOCK) (T : Mirage_time.S) (S : Mirage_stack.V4)
+    (N : Mirage_net.S) (E : Mirage_protocols.ETHERNET) (A : Mirage_protocols.ARP) (I : Mirage_protocols.IPV4) (FS: Mirage_kv.RO) = struct
+
+  module O = Openvpn_mirage.Make(R)(M)(P)(T)(S)
+
+  let read_config data =
+    FS.get data (Mirage_kv.Key.v "openvpn.config") >|= function
+    | Error e -> Rresult.R.error_to_msg ~pp_error:FS.pp_error (Error e)
+    | Ok data ->
+      let string_of_file _ = Error (`Msg "not supported") in
+      Openvpn.Config.parse_client ~string_of_file data
+
+  let local_network ip =
+    let net, my_ip = Key_gen.private_ipv4 () in
+    if Ipaddr.V4.compare my_ip ip = 0 then begin
+      Logs.warn (fun m -> m "a packet directed to us (ignoring)");
+      false
+    end else
+      Ipaddr.V4.Prefix.mem ip net
+
+  let forward_or_reject hdr payload mtu =
+    (* there are actually four potential outcomes here:
+       - packet is good to be forwarded
+       - packet gets an ICMP error report
+       - packet is a fragment and should be dropped (not handling, as noted
+         above we're reassembling)
+       - packet is an ICMP error and getting dropped (do not reply)
+    *)
+    let open Icmpv4_packet in
+    let open Icmpv4_wire in
+    let icmp_err ?(subheader = Unused) ?(code = 0) ty =
+      (* ICMP packet is 8 byte header, plus original IP header, plus original
+         payload (8 bytes) *)
+      let plen = min 8 (Cstruct.len payload) in
+      let orig_payload = Cstruct.sub payload 0 plen in
+      let ip_hdr =
+        Ipv4_packet.Marshal.make_cstruct ~payload_len:plen hdr
+      in
+      let payload = Cstruct.append ip_hdr orig_payload in
+      let icmp = { code ; ty ; subheader } in
+      let icmp_buf = Marshal.make_cstruct icmp ~payload in
+      Cstruct.append icmp_buf payload
+    in
+    let is_icmp =
+      match Ipv4_packet.Unmarshal.int_to_protocol hdr.Ipv4_packet.proto with
+      | Some `ICMP -> true
+      | _ -> false
+    and is_first_fragment = hdr.Ipv4_packet.off land 0x1FFF = 0
+    in
+    if hdr.Ipv4_packet.ttl = 1 then (* time to live exceeded *)
```

Comment:
```suggestion
    if hdr.Ipv4_packet.ttl <= 1 then (* time to live exceeded *)
```
Then reader doesn't have to go research if ip stack can accepts packets with `ttl = 0`